File: baekjoon/basic2/14502/14502.cpp
// 백준 14502 : 연구소 (https://www.acmicpc.net/problem/14502)
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
int n, m, a[10][10], visited[10][10], max_area, ny, nx;
int dy[4] = {0,0,1,-1};
int dx[4] = {1,-1,0,0};
vector<pair<int, int>> v;

void wall(int i, int j, int k, int w){
    a[v[i].first][v[i].second] = w;
    a[v[j].first][v[j].second] = w;
    a[v[k].first][v[k].second] = w;
}

void spread(int y, int x){
    visited[y][x] = 1;
    for(int i=0; i<4; i++){
        ny = y + dy[i];
        nx = x + dx[i];
        if(ny >= n || ny < 0 || nx >= m || nx < 0) continue;
        if(visited[ny][nx] == 0 && a[ny][nx] == 0) spread(ny, nx);
    }
}

int go(int y, int x){
    int ret = 1;
    visited[y][x] = 1;
    for(int i=0; i<4; i++){
        ny = y + dy[i];
        nx = x + dx[i];
        if(ny >= n || ny < 0 || nx >= m || nx < 0) continue;
        if(visited[ny][nx] == 0 && a[ny][nx] == 0) ret += go(ny, nx);
    }
    return ret;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> n >> m;
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            cin >> a[i][j];
            if(a[i][j] == 0) v.push_back({i, j});
        }
    }

    for(int i=0; i<v.size(); i++){
        for(int j=0; j<i; j++){
            for(int k=0; k<j; k++){
                wall(i, j, k, 1);
                fill(&visited[0][0], &visited[0][0] + 100, 0);
                int area = 0;

                for(int y=0; y<n; y++){
                    for(int x=0; x<m; x++){
                        if(a[y][x] == 2) spread(y, x);
                    }
                }

                for(int y=0; y<n; y++){
                    for(int x=0; x<m; x++){
                        if(visited[y][x] == 0 && a[y][x] == 0) area += go(y, x);
                    }
                }

                wall(i, j, k, 0);
                max_area = max(max_area, area);
            }
        }
    }
    cout << max_area << "\n";
    return 0;
}

File: baekjoon/basic2/14502/14502.py
# 백준 14502 : 연구소 (https://www.acmicpc.net/problem/14502)
import sys
from itertools import combinations
sys.setrecursionlimit(10**6)
dy = [0,0,1,-1]
dx = [1,-1,0,0]

def wall(yx_list, w):
    for y, x in yx_list:
        visited[y][x] = w
    return
    
def spread(y, x):
    visited[y][x] = 1
    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]
        if ny >= n or ny < 0 or nx >= m or nx < 0:
            continue
        if visited[ny][nx] == 0 and a[ny][nx] == 0:
            spread(ny, nx)
    return

def go(y, x):
    ret = 1
    visited[y][x] = 1
    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]
        if ny >= n or ny < 0 or nx >= m or nx < 0:
            continue
        if visited[ny][nx] == 0 and a[ny][nx] == 0:
            ret += go(ny, nx)
    return ret

n, m = map(int, sys.stdin.readline().split())
a = []
for i in range(n):
    a.append(list(map(int, sys.stdin.readline().split())))

wall_list = []
for i in range(n):
    for j in range(m):
        if a[i][j] == 0:
            wall_list.append((i, j))

max_area = 0
combi = list(combinations(wall_list, 3))
for com in combi:
    visited = [[0] * m for _ in range(n)]
    area = 0
    wall(com, 1)
    for i in range(n):
        for j in range(m):
            if a[i][j] == 2:
                spread(i, j)
    
    for i in range(n):
        for j in range(m):
            if visited[i][j] == 0 and a[i][j] == 0:
                area += go(i, j)
    max_area = max(max_area, area)
    wall(com, 0)

sys.stdout.write(str(max_area)+"\n")
